src/Flysystem/AssetAdapter.php

```diff
@@ -146,18 +146,29 @@ protected function configureServer($forceOverwrite = false)
 
         // Apply each configuration
         $config = new FlysystemConfig();
-        $config->set('visibility', $visibility);
         foreach ($configurations as $file => $template) {
             // Ensure file contents
             if ($forceOverwrite || !$this->has($file)) {
+                if ($this->has($file)) {
+                    $perms = $this->getVisibility($file);
+                    if ($perms['visibility'] === $visibility) {
+                        // If the visibility is already correct, do not try to re-set it.
+                        // Attempting to set the visibility of a file/folder when the filesystem is already
+                        // set correctly causes "Operation not permitted" errors on some Linux servers.
+                        $config->set('visibility', null);
+                    } else {
+                        $config->set('visibility', $visibility);
+                    }
+                }
                 // Evaluate file
                 $content = $this->renderTemplate($template);
                 $success = $this->write($file, $content, $config);
                 if (!$success) {
                     throw new Exception("Error writing server configuration file \"{$file}\"");
                 }
             }
-            $perms = $this->getVisibility($file);
+
+            $perms ??= $this->getVisibility($file);
```

Comment:
```suggestion
            $perms = $this->getVisibility($file);
```
While I get it's a tiny bit more performant, I don't like that it's getting `$perms` for something that was set two nesting levels above.  It's weird to read.

Also it's probably better to always freshly refetch the visibility in-case something went weird during $this->write() above.

Reply:
Done